[FE] Conditions List Filters

mariano-aguero opened the pull request:
Closes #408 
Closes #410 
Closes #428 
Closes #392 
Closes #430 
Closes #436 
Closes #435 

For the condition List, we can se this gif:

![Peek 08-10-2020 16-12](https://user-images.githubusercontent.com/1144028/95503112-4223b680-0981-11eb-9655-4c7819587e3d.gif)

For the prepare condition section, now we use the realitio oracle address, the input is readonly. Also fixed all the issues related to this omen prepare condition that we can see in the description
![Peek 08-10-2020 16-15](https://user-images.githubusercontent.com/1144028/95503310-929b1400-0981-11eb-99b0-25a8534df8eb.gif)

----

I added the validity filter to this issue https://github.com/gnosis/conditional-tokens-explorer/issues/409
Also, I removed the `All` option to search, because we need to improve some things in thegraph, I will create an issue for this in the thegraph repository and in this repo to improve search. At this moment we can only search for the entire hash

----

Also, I removed the option from the _dropdown actions_ called `report payout`, if the condition was already resolved

gnosis-info commented:
Travis automatic deployment:
 https://pr443--conditionaltokensexplorer.review.gnosisdev.com

mariano-aguero commented:
I added the validity filter to this issue https://github.com/gnosis/conditional-tokens-explorer/issues/409
Also, I removed the `All` option to search, because we need to improve some things in thegraph, I will create an issue for this in the thegraph repository and in this repo to improve search. At this moment we can only search for the entire hash

mariano-aguero commented:
Also, I removed the option from the _dropdown actions_ called `report payout`, if the condition was already resolved

gnosis-info commented:
Travis automatic deployment:
 https://pr443--conditionaltokensexplorer.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr443--conditionaltokensexplorer.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr443--conditionaltokensexplorer.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr443--conditionaltokensexplorer.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr443--conditionaltokensexplorer.review.gnosisdev.com

hpmaxi commented on src/components/filters/ConditionTypeFilterDropdown/index.tsx:
```diff
@@ -4,36 +4,40 @@ import { ButtonSelectLight } from 'components/buttons/ButtonSelectLight'
 interface Props {
-  onClick: (value: ConditionType | ConditionTypeAll) => void
-  value: string
+  onClick: (value: ConditionType | ConditionTypeAll, filter: Maybe<string>) => void
+  value: Maybe<string>
```
filter and value types could be more narrow?

mariano-aguero replied:
I think no

hpmaxi commented on src/components/filters/ConditionTypeFilterDropdown/index.tsx:
```diff
@@ -4,36 +4,40 @@ import { ButtonSelectLight } from 'components/buttons/ButtonSelectLight'
+  const { networkConfig } = useWeb3ConnectedOrInfura()
   const { onClick, value } = props
 
+  const oracleRealitio = networkConfig.getOracleFromName('realitio' as KnownOracle)
+
```
This could use `useMemo` with networkConfig as dependency.

mariano-aguero replied:
Done

hpmaxi commented on src/components/filters/DateFilter/index.tsx:
```diff
@@ -35,34 +37,81 @@ const FieldWrapper = styled.div`
+      onChangeFrom(event)
+    }
+
+    const fromDateMoment = moment(event.currentTarget.value)
+    const fromTimestamp = fromDateMoment.unix()
+
+    setFrom(fromTimestamp)
+  }
+
+  const onChangeToInternal = (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (typeof onChangeTo === 'function') {
+      onChangeTo(event)
+    }
+
+    const toDateMoment = moment(event.currentTarget.value)
+    const toTimestamp = toDateMoment.unix()
```
This could be changed to make all the conversion without using extra variables, `moment(...).unix()` but also moved to a helper wich make explicit in his name what you are doing, `dateToTheGraphFormat`.

mariano-aguero replied:
looking in the code, this is the only place where something similar is used. I think at the moment it doesn't make sense

hpmaxi commented on src/components/filters/DateFilter/index.tsx:
```diff
@@ -35,34 +37,81 @@ const FieldWrapper = styled.div`
+  }
+
+  // This clear the filters
+  React.useEffect(() => {
+    if (!from || !to) onSubmit(0, 0)
+  }, [from, to, onSubmit])
+
+  const errorMessage = React.useMemo(
+    () => (to && from && to < from ? 'To should be greater than From' : null),
+    [from, to]
+  )
+  const emptyValues = !from || !to
+
```
(This also applies to min, max).
emptyValues can be wrapped in `useMemo` and use in the effect leveraging morgan's law. It would be `if(!emptyValues) ... ` .

mariano-aguero replied:
Done!

gabitoesmiapodo reviewed (changes requested):
Good job, this makes using the app way easier.

❌   #410 works very well for the most part, but:

- We should do something about the Reporter / Oracle filter "Current Wallet"'s option. We could ask the user to connect, or we can disable the option. Related to this, if the user is not connected, selects "Current Wallet" and connects afterwards, no results are displayed (table is not updated).
- Number of outcomes (I know this was not specified...): If the user fills `Min` only, we should show anything above `Min`. And if they fill `Max` only, we should show anything below `Max`. It shouldn't be necessary to fill both inputs.
- The same goes for Creation Date, or any other range for that matter. 
- Hiding the filters' column should clear the filters.

Also we could probably come up with a better way to clear individual filters, though. Maybe later...

❌   #430 "Prepare" button is still enabled when there are 0 or 1 outcomes. It should be disabled.

> Expected Result: 'Prepare' button is disabled. user is not allowed to create a condition with 1 outcome

I added a note below the outcomes' list to make it more clear.

Also, we should make the min / max outcomes part of the `.env` vars (they're hardcoded in a few places now).

❌  #435 This works fine, but:

"Question must not exist" is not clear enough, the user will probably not know what we're talking about.

Probably we should use the warning component (see "Merge Positions" when using a resolved condition), something like:

**Title:** "Duplicated Question"
**Text:** "The question for this condition already exists on [ORACLE HERE]. Please change it to a different one."

🆗  #408
🆗  #428
🆗  #392
🆗  #436

hpmaxi commented on src/config/networkConfig.ts:
```diff
@@ -365,6 +365,23 @@ export class NetworkConfig {
   }
 
+  getOracleFromName(oracleName: KnownOracle): Oracle {
+    const oracles = networks[this.networkId].oracles
+
+    for (const oracle of oracles) {
+      if (oracle.name === oracleName) {
+        return oracle
+      }
+    }
+
+    return {
+      name: 'unknown',
+      description: 'Unknown',
+      url: '',
+      address: '',
+    }
+  }
+
```
If unknown oracle is named `unknownOracle` thsi could also be written like
`networks[this.networkId].oracles.find(o => oracle.name === oracleName) || unknownOracle`, to me is shorter and clearer but is a matter of taste, so take it with a grain of salt

mariano-aguero replied:
Mmm I will not do this, as you say is a matter of taste

gnosis-info commented:
Travis automatic deployment:
 https://pr443--conditionaltokensexplorer.review.gnosisdev.com

hpmaxi commented on src/hooks/useConditionsList.tsx:
```diff
@@ -0,0 +1,61 @@
+
+  const query = buildQueryConditionsList(advancedFilter)
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const variables: { [k: string]: any } = {}
+  if (ReporterOracle.type === OracleFilterOptions.Custom)
+    variables['oracleNotIn'] = ReporterOracle.value
+  if (
```
There must be a better way to build queries and variables, but for the moment this should be accompanied with a test at least. About the type of variables, we can use `[k in someType]: anotherType` , if both types are generic we have a `Record`, I found that it is better to have only some valid keys in variables.

mariano-aguero replied:
I didn't find any other solution or a queryBuilder or a solution that the library gives me.

gnosis-info commented:
Travis automatic deployment:
 https://pr443--conditionaltokensexplorer.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr443--conditionaltokensexplorer.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr443--conditionaltokensexplorer.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr443--conditionaltokensexplorer.review.gnosisdev.com

mariano-aguero commented:
> Good job, this makes using the app way easier.
> 
>  #410 works very well for the most part, but:
> 
> * We should do something about the Reporter / Oracle filter "Current Wallet"'s option. We could ask the user to connect, or we can disable the option. Related to this, if the user is not connected, selects "Current Wallet" and connects afterwards, no results are displayed (table is not updated).
> * Number of outcomes (I know this was not specified...): If the user fills `Min` only, we should show anything above `Min`. And if they fill `Max` only, we should show anything below `Max`. It shouldn't be necessary to fill both inputs.
> * The same goes for Creation Date, or any other range for that matter.
> * Hiding the filters' column should clear the filters.
> 
> Also we could probably come up with a better way to clear individual filters, though. Maybe later...
> 
>  #430 "Prepare" button is still enabled when there are 0 or 1 outcomes. It should be disabled.
> 
> > Expected Result: 'Prepare' button is disabled. user is not allowed to create a condition with 1 outcome
> 
> I added a note below the outcomes' list to make it more clear.
> 
> Also, we should make the min / max outcomes part of the `.env` vars (they're hardcoded in a few places now).
> 
>  #435 This works fine, but:
> 
> "Question must not exist" is not clear enough, the user will probably not know what we're talking about.
> 
> Probably we should use the warning component (see "Merge Positions" when using a resolved condition), something like:
> 
> **Title:** "Duplicated Question"
> **Text:** "The question for this condition already exists on [ORACLE HERE]. Please change it to a different one."
> 
>  #408
>  #428
>  #392
>  #436

Nice feedback! 

I finish with everything

gnosis-info commented:
Travis automatic deployment:
 https://pr443--conditionaltokensexplorer.review.gnosisdev.com

elena-zh commented:
Great changes! 
I also I have reviewed the implemented features and have some notes (besides the listed above): 
1. Once a Redeem position option is removed for a resolved condition in the list, it would be nice to remove is from a position details as well
2. #408: Search returns an odd error when search by QuesitonID/Oracle Address/Creator Address and fill in the search field with numbers
![error](https://user-images.githubusercontent.com/70885163/95754447-b6c16280-0cab-11eb-875d-4939057b38b5.jpg)
3. #410: GraphQL fails when I perform 3 or more search queries =)
![3-4 seqrch queries](https://user-images.githubusercontent.com/70885163/95754493-cccf2300-0cab-11eb-8461-a07222ff37ae.jpg)
4. #410: Column names overlap each other when filter is opened
![opened filters](https://user-images.githubusercontent.com/70885163/95754714-1ddf1700-0cac-11eb-92bd-0cea9001b20c.jpg)
5. #430: It would be nice to show a warning message after filling in the 1st outcome rather then after filling in the Question field (from me perspective, every red message looks a bit scaring for a user, especially when a form is not completely filled it)
6. #435 : still do not see any warning message, even if I create the same condition within an hour. =(
7. #392 : looks like a resolution date is completely removed from a condition details. is it an expected behavior?


Look good: 
#436 
#428 

Also, a lot of the issues reported in #440 are fixed (marked them as fixed there). However, several ones are still unfixed, I made them in bold. It would be nice to take a look into them as well. 

Thanks a lot!

gabitoesmiapodo reviewed (approved):
- 🆗  We should do something about the Reporter / Oracle filter "Current Wallet"'s option (...) -> There's a small problem with this, I created an issue here: 
- 🆗  Number of outcomes (I know this was not specified...) (...) -> Added `onPressEnter` to this.
- 🆗  The same goes for Creation Date, or any other range for that matter. (...) -> Added `onPressEnter` to this.
- 🆗  Hiding the filters' column should clear the filters.
- 🆗  #430 "Prepare" button is still enabled when there are 0 or 1 outcomes. It should be disabled. (...) -> I removed the error message, seems redundant when having a note stating the rules for this.
- 🆗  Also, we should make the min / max outcomes part of the .env vars (...)
- 🆗  "Question must not exist" is not clear enough, (...)

mariano-aguero commented:
> Great changes!
> I also have reviewed the implemented features and have some notes (besides the listed above):
> 
> 1. Once a Redeem position option is removed for a resolved condition in the list, it would be nice to remove it from a position details as well

I removed the report payout option, not the redeem option. I opened an issue for this https://github.com/gnosis/conditional-tokens-explorer/issues/458


> 4. #410: Column names overlap each other when filter is opened
>    ![opened filters](https://user-images.githubusercontent.com/70885163/95754714-1ddf1700-0cac-11eb-92bd-0cea9001b20c.jpg)

Created an issue for this https://github.com/gnosis/conditional-tokens-explorer/issues/459

> 7. #392 : looks like a resolution date is completely removed from a condition details. is it an expected behavior?

Yes, is remove it when the condition is not resolved, because there is no timestamp

mariano-aguero commented:
@elena-zh can you create issues for every error that you catch ? I really appreciate it 
If you can't, please let me know

elena-zh commented:
@mariano-aguero , tasks for pp.1, 4 ,7 were created by you, thanks. 
I have create a separate ticket for p.2 - #467, for p.6 - reopened #435.
pp. 3, 5 are not reproducible so far.
